### ISSUE TITLE
Non-streaming decode() appears to remove the BOM?  

### ISSUE THREAD

**Issue by dralley** [hsivonen/encoding_rs#88]
https://github.com/hsivonen/encoding_rs/blob/d4d7d2a99aac266ecf6938c3832aefaaf8c1e52b/src/lib.rs#L2974-L2980

Functionally, `decode()` and `decode_with_bom_removal()` seem pretty much the same?  That doesn't seem correct?  If there's a variant called "decode_with_bom_removal" then I would expect the standard variant *not* to do this.

It's totally valid to decode the BOM, the BOM a unicode character like any other character.

```Rust
    pub fn decode<'a>(&'static self, bytes: &'a [u8]) -> (Cow<'a, str>, &'static Encoding, bool) {
        let (encoding, without_bom) = match Encoding::for_bom(bytes) {
            Some((encoding, bom_length)) => (encoding, &bytes[bom_length..]),
            None => (self, bytes),
        };
        let (cow, had_errors) = encoding.decode_without_bom_handling(without_bom);   // <--- it's decoding and returning only the part without the bom???
        (cow, encoding, had_errors)
    }

    pub fn decode_with_bom_removal<'a>(&'static self, bytes: &'a [u8]) -> (Cow<'a, str>, bool) {
        let without_bom = if self == UTF_8 && bytes.starts_with(b"\xEF\xBB\xBF") {
            &bytes[3..]
        } else if (self == UTF_16LE && bytes.starts_with(b"\xFF\xFE"))
            || (self == UTF_16BE && bytes.starts_with(b"\xFE\xFF"))
        {
            &bytes[2..]
        } else {
            bytes
        };   // <--- This is all essentially the same thing that `Encoding::for_bom()` does
        self.decode_without_bom_handling(without_bom)  // <--- Using "self" instead of a detected encoding is slightly different
    }

```

```Rust
use encoding_rs::*;

fn main() {
    // Two characters, '1' and then BOM character
    println!("{:?}", UTF_16LE.decode(&[0x31, 0x00, 0xFF, 0xFE])); 
    // Nothing - BOM removed
    println!("{:?}", UTF_16LE.decode(&[0xFF, 0xFE]));
}
```
```
("1\u{feff}", Encoding { UTF-16LE }, false)
("", Encoding { UTF-16LE }, false)
```

**Comment by dralley** [comment]
The same thing happens with the streaming API.  Am I missing something?

```
use encoding_rs::*; // 0.8.31

fn main() {
    // // Two characters, '1' and then BOM character
    // println!("{:?}", UTF_16LE.decode(&[0x31, 0x00, 0xFF, 0xFE]).0.as_bytes()); 
    // // Nothing - BOM removed
    // println!("{:?}", UTF_16LE.decode(&[0xFF, 0xFE]).0.as_bytes());
    
    let mut buf = [0; 100];
    
    dbg!(UTF_16LE.new_decoder().decode_to_utf8(
        &[0x31, 0x00, 0xFF, 0xFE],
        &mut buf,
        true,
    ));
    
    println!("{:?}", &buf);
    
    let mut buf = [0; 100];
    
    dbg!(UTF_16LE.new_decoder().decode_to_utf8(
        &[0xFF, 0xFE],
        &mut buf,
        true,
    ));
    
    println!("{:?}", &buf);
}
```

**Comment by dralley** [comment]
@hsivonen Is there a detail that I'm missing?

**Comment by dralley** [comment]
I could try to put up a PR, I just want to verify that there's not a reason behind this

**Comment by dralley** [comment]
@hsivonen That seems problematic - there is no way to respect the BOM (use the BOM-indicated encoding) without removing it.  I want the `decode()` behavior but I don't want to remove the BOM, and I don't see why the BOM should be removed by default.  It's valid unicode, and sometimes you do want it.

The name of `decode_with_bom_removal()` doesn't indicate the actual difference between `decode_with_bom_removal()` and `decode()`.

**Comment by dralley** [comment]
Furthermore I don't see any documentation indicating that `decode()` removes the BOM, could you point out where it is documented?

**Comment by hsivonen** [comment]
> Furthermore I don't see any documentation indicating that `decode()` removes the BOM, could you point out where it is documented?

Good point. It says "BOM sniffing" without saying explicitly that BOM sniffing means BOM removal.

What's the use case for deciding the encoding from BOM but still letting the BOM show through to output?

**Comment by dralley** [comment]
I'm working on improving the encoding support of `quick-xml`.  One idea I had is that a user may want to get either:

* a normalized series of XML events, which strips any non-meaningful whitespace including anything prior to the XML delcaration (including the BOM), or
* a  series of XML events, which would include whitespace between elements (and the BOM), which if passed back into a writer would exactly reproduce the document

The latter would potentially simplify making edits to existing documents, given that some software wants to see a UTF-8 BOM even though the spec recommends against it.

In order to pull that off, the BOM would need to not be stripped automatically.  Whether it gets removed or not (alongside other things) would be handled by the Reader.

**Comment by hsivonen** [comment]
The BOM isn't part of the XML information set, so an XML use case hasn't arisen before. Does you exact reproduction mode also retain whitespace on either side of the equals sign for attributes, etc.?

Since this issue is filed against the non-streaming API, you can easily find out if there was a BOM by inspecting the buffer with [`Encoding::for_bom()`](https://docs.rs/encoding_rs/latest/encoding_rs/struct.Encoding.html#method.for_bom).

**Comment by hsivonen** [comment]
> Well, the streaming API does the same thing as mentioned in the first comment after the OP, I just first encountered this while doing some experimentation. The streaming API is what we would be using in practice.

I see. Still, I think at least for now, I'm going to continue to treat this as out of scope, since the BOM with BOM semantics isn't supposed to be part of the logical content of the stream.

**Comment by dralley** [comment]
>the BOM with BOM semantics isn't supposed to be part of the logical content of the stream.

This is the part I'm unsure on.  Nothing about the BOM is particularly special, it's just a unicode character, etc.  Isn't it just part of the document, then?  More akin to the `#!` shebang line at the beginning of a Python/Perl/Bash script than magic bytes at the beginning of a zip file, except that Unicode says it should present as an invisible zero-width non-breaking space.

Various documents I've read don't spell it out explicitly but they make it seem like part of the data stream itself.

https://www.w3.org/International/questions/qa-byte-order-mark
https://unicode.org/faq/utf_bom.html#bom1

**Comment by hsivonen** [comment]
See [The Unicode Standard 15.0](https://www.unicode.org/versions/Unicode15.0.0/UnicodeStandard-15.0.pdf) D98 second bullet last sentence on page 131 (PDF page 157). See also the note in [The Encoding Standard](https://encoding.spec.whatwg.org/#encodings) that explains that the naming doesn't match the naming in The Unicode Standard.

Note that the XML spec says of the BOM: "This is an encoding signature, not part of either the markup or the character data of the XML document." Also, the BNF in the XML spec doesn't consider the BOM part of the textual syntax.

**Comment by dralley** [comment]
That appears to be correct for UTF-16.  Regarding UTF-8, the bullet points under D95 on the previous page are frustratingly unclear - it has no such language and actually seems to imply the opposite.

>While there is obviously no need for a byte order signature when using UTF-8,
there are occasions when processes convert UTF-16 or UTF-32 data contain-
ing a byte order mark into UTF-8. When represented in UTF-8, the byte order
mark turns into the byte sequence <EF BB BF>. Its usage at the beginning of a
UTF-8 data stream is neither required nor recommended by the Unicode Stan-
dard, but its presence does not affect conformance to the UTF-8 encoding
scheme. Identification of the <EF BB BF> byte sequence at the beginning of a
data stream can, however, be taken as a near-certain indication that the data
stream is using the UTF-8 encoding scheme.

But considering I really don't feel like lawyering the spec to that degree, point taken, we should just handle the BOM separately.
